### DISCUSSION TITLE
Ho to call an object method with another object method as argument?

### DISCUSSION THREAD

**Discussion by SamueleD78** [vfsfitvnm/frida-il2cpp-bridge#153] [Q&A]
So i have a method in an object

`    System.Void ForEach(System.Func<Nabe.Internal.File,System.Boolean> func); // 0x02be142c
`
and i have another method in another object with the same params:

`    System.Boolean IsExistOnLocal(Nabe.Internal.File file, System.Boolean isStrict); // 0x02fff828
`

i could get the reference to the methods:

```
const IsExistOnLocal = inst.method<boolean>("IsExistOnLocal");

const Nabe_Internal_Catalog = new Il2Cpp.Object(inst.method<Il2Cpp.Pointer>("get_Catalog").invoke());
const Nabe_Internal_Catalog_ForEach = Nabe_Internal_Catalog.method("ForEach");

```

now, how can i call `Nabe_Internal_Catalog_ForEach` passing `IsExistOnLocal` as argument?

i've tryed hard without understanding how ^^''
ty so much in advance

**Comment by vfsfitvnm** [comment]
Well, this is a problematic territory. You should find a way to create a `System.Func<Nabe.Internal.File, System.Boolean>` object.

This is its constructor:
```cs
System.Void .ctor(System.Object object, System.IntPtr method);
```
`object` should be the instance the method will be called against, `method` is a pointer (should be the one you get via `handle`, not `virtualAddress`).

You may take a shot at it.

PS - This line
```ts
const Nabe_Internal_Catalog = new Il2Cpp.Object(inst.method<Il2Cpp.Pointer>("get_Catalog").invoke());
```
is "incorrect" (it works anyway because of abstractions), you should do:
```ts
// prefer lowerCase for C# objects to avoid confusion with methods/classes
const nabeInternalCatalog = inst.method<Il2Cpp.Object>("get_Catalog").invoke();
```

**Reply by SamueleD78** [reply]
Thank you for the suggestions! Yes my bad, too many copy & paste

Mmh i don't understand why i have to create the object.
the `IsExistOnLocal` it's exactly the function needed, and it already exists in an object.

if it was normal code, i woul probably had called this as:

`objA.forEach(objB.IsExistOnLocal);`

i've tryed

`ForEach.invoke(IsExistOnLocal.handle);`

but it gives an access violation

**Reply by vfsfitvnm** [reply]
> Mmh i don't understand why i have to create the object.

You have to because it expects a C# object.

`IsExistOnLocal.handle` returns a pointer to a `Il2Cpp.Method`, a native struct, (which is not a `System.Reflection.MethodInfo` nor a `System.Func`).

**Reply by SamueleD78** [reply]
understood, my mistake in messing System.Func with a normal function and not an object. So that would be:

```const test = Il2Cpp.Image.corlib.class("System.Func`1").inflate(Il2Cpp.Image.corlib.class("System.Boolean"));```

result:
```
class System.Func<System.Boolean> : System.MulticastDelegate
{
    
    System.Void .ctor(System.Object object, System.IntPtr method); // 0x02ec0a64
    System.Boolean Invoke(); // 0x02d975dc
    System.IAsyncResult BeginInvoke(System.AsyncCallback callback, System.Object object); // 0x02322560
    System.Boolean EndInvoke(System.IAsyncResult result); // 0x0272294c
}

```
then:

```
const test2 = test.alloc();
test2.method(".ctor").invoke(inst, IsExistOnLocal.handle);
Internal_Catalog_ForEach.invoke(test2);
```

But access violation again. What i'm doing wrong?

**Reply by vfsfitvnm** [reply]
You are obtaining a `System.Func<System.Boolean>`, but a `System.Func<Nabe.Internal.File, System.Boolean>` is required:
```ts
const SystemBoolean = Il2Cpp.Image.corlib.class("System.Boolean");
const SystemFunc = Il2Cpp.Image.corlib.class("System.Func`2").inflate(/**/, SystemBoolean);
```
Alternatively, you could just do:
```ts
const SystemFunc = nabeInternalCatalog.method("ForEach").parameter("func").type.class;
```

**Reply by SamueleD78** [reply]
Ohhhh now i get it!!! Thank you so much!!

Being a Func, the first param is a calling param (T), the last param it's always the returning value.

So actually i needed to search for a `boolean func(struct)`, not for a `void func(struct, bool)`.
I did  it following your advices and it's working flawlessly.

Thank you SO much, you really helped me a lot ^^
